cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -80,13 +91,114 @@ func ValidateTokenDiscovery(c *kubeadm.TokenDiscovery, fldPath *field.Path) fiel
 	return allErrs
 }
 
+func ValidateDiscovery(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateTLSBootstrapToken(c, fldPath)...)
+	var count int
+	if len(c.DiscoveryToken) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryToken(c, fldPath)...)
+		count++
+	}
+	if len(c.DiscoveryFile) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryFile(c, fldPath)...)
+		count++
+	}
+	if len(c.DiscoveryURL) != 0 {
+		allErrs = append(allErrs, ValidateDiscoveryURL(c, fldPath)...)
+		count++
+	}
+	if count != 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one discovery strategy can be provided"))
+	}
+	return allErrs
+}
+
+func ValidateTLSBootstrapToken(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.Token) != 0 && (len(cfg.TLSBootstrapToken) != 0 || len(cfg.DiscoveryToken) != 0) {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "--token are mutually exclusive with --bootstrap-token and --discovery-token"))
+	}
+	if len(cfg.Token) == 0 && (len(cfg.TLSBootstrapToken) == 0 || len(cfg.DiscoveryToken) == 0) {
```

Comment:
Not sure how to capture this, but if you set `discovery-file` or `discovery-url` then you don't need `discovery-token`.

In other words, `kubeadm join --discovery-url=... --tls-bootstrap-token=...` should work but I think that it won't work in this case.

One way we could make this work is to make the more specific token override the generic token.

So -- `kubeadm join --token=AAA --discovery-token=BBB` would mean that the discovery token is BBB and the tls bootstrap token is AAA.

Similarly, `kubeadm join --discovery-url=http://... --token=AAA` would mean draw discovery from a URL (no token needed) and the tls bootstrap token is AAA.